// @sinclair/typebox/value/pointer.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox/value

The MIT License (MIT)

Copyright (c) 2022 Haydn Paterson (sinclair) <haydn.developer@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValuePointer = exports.ValuePointerRootDeleteError = exports.ValuePointerRootSetError = void 0;
class ValuePointerRootSetError extends Error {
    constructor(value, path, update) {
        super('ValuePointer: Cannot set root value');
        this.value = value;
        this.path = path;
        this.update = update;
    }
}
exports.ValuePointerRootSetError = ValuePointerRootSetError;
class ValuePointerRootDeleteError extends Error {
    constructor(value, path) {
        super('ValuePointer: Cannot delete root value');
        this.value = value;
        this.path = path;
    }
}
exports.ValuePointerRootDeleteError = ValuePointerRootDeleteError;
/** ValuePointer performs mutable operations on values using RFC6901 Json Pointers */
var ValuePointer;
(function (ValuePointer) {
    /** Formats the path into navigable components */
    function* Format(path) {
        function clear(chars) {
            while (chars.length > 0)
                chars.shift();
        }
        const chars = [];
        for (let i = 0; i < path.length; i++) {
            const char = path.charAt(i);
            if (char === '/') {
                if (i !== 0) {
                    yield chars.join('');
                    clear(chars);
                }
            }
            else if (char === '~' && path.charAt(i + 1) === '0' && (path.charAt(i + 2) === '/' || i !== path.length - 1)) {
                chars.push('~');
                i += 1;
            }
            else if (char === '~' && path.charAt(i + 1) === '1' && (path.charAt(i + 2) === '/' || i !== path.length - 1)) {
                chars.push('/');
                i += 1;
            }
            else {
                chars.push(char);
            }
        }
        yield chars.join('');
        clear(chars);
    }
    /** Sets the value at the given pointer. If the value at the pointer does not exist it is created. */
    function Set(value, path, update) {
        if (path === '')
            throw new ValuePointerRootSetError(value, path, update);
        const pointer = [...Format(path)];
        let current = value;
        while (pointer.length > 1) {
            const next = pointer.shift();
            if (current[next] === undefined)
                current[next] = {};
            current = current[next];
        }
        current[pointer.shift()] = update;
    }
    ValuePointer.Set = Set;
    /** Deletes a value at the given path. */
    function Delete(value, path) {
        if (path === '')
            throw new ValuePointerRootDeleteError(value, path);
        let current = value;
        const pointer = [...Format(path)];
        while (pointer.length > 1) {
            const next = pointer.shift();
            if (current[next] === undefined)
                return;
            current = current[next];
        }
        if (globalThis.Array.isArray(current)) {
            const index = parseInt(pointer.shift());
            return current.splice(index, 1);
        }
        else {
            const key = pointer.shift();
            delete current[key];
        }
    }
    ValuePointer.Delete = Delete;
    /** True if a value exists at the given path */
    function Has(value, path) {
        if (path === '')
            return true;
        let current = value;
        const pointer = [...Format(path)];
        while (pointer.length > 1) {
            const next = pointer.shift();
            if (current[next] === undefined)
                return false;
            current = current[next];
        }
        return current[pointer.shift()] !== undefined;
    }
    ValuePointer.Has = Has;
    /** Gets the value at the given path */
    function Get(value, path) {
        if (path === '')
            return value;
        let current = value;
        const pointer = [...Format(path)];
        while (pointer.length > 1) {
            const next = pointer.shift();
            if (current[next] === undefined)
                return undefined;
            current = current[next];
        }
        return current[pointer.shift()];
    }
    ValuePointer.Get = Get;
})(ValuePointer = exports.ValuePointer || (exports.ValuePointer = {}));
