// @microsoft/fast-element/dist/esm/index.js
export * from "./platform.js";
export * from "./templating/template.js";
export * from "./components/fast-element.js";
export { FASTElementDefinition, } from "./components/fast-definitions.js";
export * from "./components/attributes.js";
export * from "./components/controller.js";
export * from "./templating/compiler.js";
export { ElementStyles, } from "./styles/element-styles.js";
export { css, cssPartial } from "./styles/css.js";
export { CSSDirective } from "./styles/css-directive.js";
export * from "./templating/view.js";
export * from "./observation/observable.js";
export * from "./observation/notifier.js";
export {} from "./observation/array-change-records.js";
export { enableArrayObservation } from "./observation/array-observer.js";
export { DOM } from "./dom.js";
export * from "./templating/binding.js";
export * from "./templating/html-directive.js";
export * from "./templating/ref.js";
export * from "./templating/when.js";
export * from "./templating/repeat.js";
export * from "./templating/slotted.js";
export * from "./templating/children.js";
export { elements, } from "./templating/node-observation.js";
